mediatek: filogic: add support for Xiaomi WR30U

hank9999 opened the pull request:
Hardware specification:
  SoC: MediaTek MT7981B 2x A53
  Flash: ESMT F50L1G41LB 128MB
  RAM: NT52B128M16JR-FL 256MB
  Ethernet: 4x 10/100/1000 Mbps
  Switch: MediaTek MT7531AE
  WiFi: MediaTek MT7976C
  Button: Reset, Mesh
  Power: DC 12V 1A

Flash instructions:

1. Get ssh access
   Check this link: https://forum.openwrt.org/t/openwrt-support-for-xiaomi-ax3000ne/153769/22

2. Backup import partitions
   ```
   dev:    size   erasesize  name
   mtd1: 00100000 00020000 "BL2"
   mtd2: 00040000 00020000 "Nvram"
   mtd3: 00040000 00020000 "Bdata"
   mtd4: 00200000 00020000 "Factory"
   mtd5: 00200000 00020000 "FIP"
   mtd8: 02200000 00020000 "ubi"
   mtd9: 02200000 00020000 "ubi1"
   mtd12: 00040000 00020000 "KF"

   ```

   Use these commands blow to backup your stock partitions.
   ```
   nanddump -f /tmp/BL2.bin /dev/mtd1
   nanddump -f /tmp/Nvram.bin /dev/mtd2
   nanddump -f /tmp/Bdata.bin /dev/mtd3
   nanddump -f /tmp/Factory.bin /dev/mtd4
   nanddump -f /tmp/FIP.bin /dev/mtd5
   nanddump -f /tmp/ubi.bin /dev/mtd8
   nanddump -f /tmp/KF.bin /dev/mtd12
   ```
   Then, transfer them to your computer via scp, netcat, tftp or others and keep them in a safe place.

3. Setup Nvram
   Get the current stock: `cat /proc/cmdline`

   If you find `firmware=0` or `mtd=ubi`, use these commands:
   ```
   nvram set boot_wait=on
   nvram set uart_en=1
   nvram set flag_boot_rootfs=1
   nvram set flag_last_success=1
   nvram set flag_boot_success=1
   nvram set flag_try_sys1_failed=0
   nvram set flag_try_sys2_failed=0
   nvram commit
   ```
   
   If you find `firmware=1` or `mtd=ubi1`, use these commands:
   ```
   nvram set boot_wait=on
   nvram set uart_en=1
   nvram set flag_boot_rootfs=0
   nvram set flag_last_success=0
   nvram set flag_boot_success=1
   nvram set flag_try_sys1_failed=0
   nvram set flag_try_sys2_failed=0
   nvram commit
   ```
4. Flash stock-initramfs-factory.ubi
   If you find `firmware=0` or `mtd=ubi`:
   `ubiformat /dev/mtd9 -y -f /tmp/stock-initramfs-factory.ubi`

   If you find `firmware=1` or `mtd=ubi1`:
   `ubiformat /dev/mtd8 -y -f /tmp/stock-initramfs-factory.ubi`

   Then reboot your router, it should boot to the openwrt initramfs system now.

5. Setup uboot-env
   Now it will be setup automatically in upgrade process, you can skip this step.

   If your `fw_setenv` did not work, you need run this command:
   `echo "/dev/mtd1 0x0 0x10000 0x20000" > /etc/fw_env.config`

   Then setup uboot-env:
   ```
   fw_setenv boot_wait on
   fw_setenv uart_en 1
   fw_setenv flag_boot_rootfs 0
   fw_setenv flag_last_success 1
   fw_setenv flag_boot_success 1
   fw_setenv flag_try_sys1_failed 8
   fw_setenv flag_try_sys2_failed 8
   fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),34816k(ubi),34816k(ubi1),32768k(overlay),12288k(data),256k(KF)"
   ```

6. Flash stock-squashfs-sysupgrade.bin
   Use shell command: 
   `sysupgrade -n /tmp/stock-squashfs-sysupgrade.bin` 
   Or go to luci web.


If you need to change to Openwrt U-Boot layout, do next. If you do not need, please ignore it.

Change to OpenWrt U-Boot:

1. Flash ubootmod-initramfs-factory.ubi
   Check mtd partitions: `cat /proc/mtd`
   ```
   mtd7: 00040000 00020000 "KF"
   mtd8: 02200000 00020000 "ubi_kernel"
   mtd9: 04e00000 00020000 "ubi"
   ```

   Run following command: 
   `ubiformat /dev/mtd8 -y -f /tmp/ubootmod-initramfs-factory.ubi`
   Then reboot your router, it should boot to the openwrt initramfs system now.

2. Check mtd again
   ```
   mtd7: 00040000 00020000 "KF"
   mtd8: 07000000 00020000 "ubi"
   ```
   Make sure mtd8 is ubi.

3. Install kmod-mtd-rw
   Run command: `opkg update && opkg install kmod-mtd-rw`
   Or get it in openwrt server, or build it yourself, then install it manually

   Then run this command: 
   `insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1`

4. Clean up pstore
   Run Command: `rm -f /sys/fs/pstore/*`

5. Format ubi and create new ubootenv volume
   ```
   ubidetach -p /dev/mtd8; ubiformat /dev/mtd8 -y; ubiattach -p /dev/mtd8
   ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
   ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB
   ```

6. (Optional) Add recovery boot feature.
   ```
   ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
   ubiupdatevol /dev/ubi0_2 /tmp/ubootmod-initramfs-recovery.itb
   ```

7. Flash Openwrt U-Boot
   ```
   mtd write /tmp/ubootmod-preloader.bin BL2
   mtd write /tmp/ubootmod-bl31-uboot.fip FIP
   ```

6. Flash ubootmod-squashfs-sysupgrade.itb
   Use shell command: 
   `sysupgrade -n /tmp/ubootmod-squashfs-sysupgrade.itb` 
   Or go to luci web.

Now everything is done, Enjoy!


Go Back to stock from Openwrt U-Boot:

1. Force flash ubootmod-initramfs-recovery.itb
   Use shell command: 
   `sysupgrade -F -n /tmp/ubootmod-initramfs-recovery.itb` 
   Or go to luci web.
   Then it should boot to the openwrt initramfs system now.

2. Format ubi and Nvram
   ```
   ubidetach -p /dev/mtd8; ubiformat /dev/mtd8 -y; ubiattach -p /dev/mtd8
   mtd erase Nvram
   ```

3. Install kmod-mtd-rw
   Run command: `opkg update && opkg install kmod-mtd-rw`
   Or get it in openwrt server, or build it yourself, then install it manually

   Then run this command: 
   `insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1`

4. Flash stock U-Boot and ubi
   ```
   mtd write /tmp/BL2.bin BL2
   mtd write /tmp/FIP.bin FIP
   mtd write /tmp/ubi.bin ubi
   ```
   Then reboot your router, waiting it finished rollback in minutes.


Go Back to stock from stock layout Openwrt:
   Just run command: `ubiformat /dev/mtd8 -y -f /tmp/ubi.bin`
   Then reboot your router, waiting it finished rollback in minutes.


Notes:
1. Openwrt U-Boot and ubootmod openwrt did not enable NMBM.
   Please make your backup safe.

aiamadeus commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -37,10 +37,12 @@ bananapi,bpi-r3)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x10000" "0x20000"
 	ubootenv_add_uci_sys_config "/dev/mtd2" "0x0" "0x10000" "0x20000"
 	;;
+xiaomi,mi-router-wr30u-ubootmod|\
```
The order of letters should be kept.

hank9999 replied:
Corrected.

aiamadeus commented on package/boot/uboot-mediatek/Makefile:
```diff
@@ -314,6 +326,7 @@ UBOOT_TARGETS := \
 	ravpower_rp-wd009 \
 	mt7629_rfb \
 	mt7981_qihoo_360t7 \
+    mt7981_xiaomi_mi-router-wr30u \
```
Use Tab here.

hank9999 replied:
Corrected.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u.dtsi:
```diff
@@ -0,0 +1,236 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=8 earlycon=uart8250,mmio32,0x11002000";
```
`stdout-path = "serial0:115200n8";` That's enough.

hank9999 replied:
Corrected.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u.dtsi:
```diff
@@ -0,0 +1,236 @@
+		};
+
+		mesh {
+			label = "mesh";
```
Add: `linux,input-type = <EV_SW>;`

hank9999 replied:
Corrected. And, I also add it to u-boot dts.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u.dtsi:
```diff
@@ -0,0 +1,236 @@
+		compatible = "mediatek,eth-mac";
+		reg = <0>;
+		phy-mode = "2500base-x";
+
```
Set lan mac here:
```
		nvmem-cells = <&macaddr_factory_4>;
		nvmem-cell-names = "mac-address";
		mac-address-increment = <(-1)>;
```

hank9999 replied:
Corrected. I recheck stock firmware, it should be this, my understand before is wrong, thank you.

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -71,9 +75,12 @@ mediatek_setup_macs()
 	xiaomi,redmi-router-ax6000-stock|\
 	xiaomi,redmi-router-ax6000-ubootmod)
 		wan_mac=$(mtd_get_mac_ascii Bdata ethaddr_wan)
+		lan_mac=$wan_mac
```
Looks wrong.

hank9999 replied:
Corrected. I recheck stock firmware, it should be removed, my understand before is wrong, thank you.

hank9999 replied:
> may be we can fix it ,wr30u mac from mtd3(Bdata), ethaddr_wan is mac ,ethaddr is 5G mac We can refer to this [padavanonly/immortalwrtARM@f76f9fc](https://github.com/padavanonly/immortalwrtARM/commit/f76f9fc1baf2e0fd502611f86d4ca23420fe1248) wan_mac=$(dd if=/dev/mtd3ro bs=1 count=256 2>/dev/null |grep "^ethaddr_wan=" | awk -F '=' '{print $2}') lan_mac=$(dd if=/dev/mtd3ro bs=1 count=256 2>/dev/null |grep "^ethaddr=" | awk -F '=' '{print $2}')



> may be we can fix it ,wr30u mac from mtd3(Bdata), ethaddr_wan is mac ,ethaddr is 5G mac We can refer to this [padavanonly/immortalwrtARM@f76f9fc](https://github.com/padavanonly/immortalwrtARM/commit/f76f9fc1baf2e0fd502611f86d4ca23420fe1248) wan_mac=$(dd if=/dev/mtd3ro bs=1 count=256 2>/dev/null |grep "^ethaddr_wan=" | awk -F '=' '{print $2}') lan_mac=$(dd if=/dev/mtd3ro bs=1 count=256 2>/dev/null |grep "^ethaddr=" | awk -F '=' '{print $2}')

It is unneeded. Everything works fine. WIFI, Lan and Wan MACs are all correct.

aiamadeus commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -70,6 +79,7 @@ platform_do_upgrade() {
 	cudy,wr3000-v1)
 		default_do_upgrade "$1"
 		;;
+	xiaomi,mi-router-wr30u-ubootmod|\
```
The order of letters should be kept.

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -257,6 +257,51 @@ define Device/qihoo_360t7
 endef
 TARGET_DEVICES += qihoo_360t7
 
+define Device/xiaomi_mi-router-wr30u-stock
```
Order by the letter...

hank9999 commented:
Sorry for force push same changes again, because of some mistake in the commit message.

ptpt52 reviewed (commented):
LGTM

abandon-xyz commented:
LGTM

liudf0716 commented:
mark

Aabandon reviewed (commented):
good job

iamyangyiok commented:
good job

PatriciaLee3 commented:
It works fine on my device.

BTW: You need to change your commit information using your real name to make it possible to be merged.

hank9999 commented:
> It works fine on my device.
> 
> BTW: You need to change your commit information using your real name to make it possible to be merged.

Thanks for your tip, I will fix it.

hank9999 commented:
Updates:
1. Resolve some conflicts
2. Remove some extra spaces
3. Submit with my real English name

hank9999 committed:
uboot-mediatek: add support for Xiaomi WR30U

Add a custom uboot build to support openwrt uboot layout.

Signed-off-by: Hank Moretti <mchank9999@gmail.com>

hank9999 commented on package/boot/uboot-mediatek/patches/434-add-xiaomi_mi-router-wr30u.patch:
```diff
@@ -0,0 +1,460 @@
++		reg = <0>;
++		spi-max-frequency = <52000000>;
++
++		mediatek,nmbm;
```
I forgot to remove this code after testing, it have been corrected, thanks.

By the way, i think it's important to have an openwrt firmware with NMBM as a choice, because there are indeed many people using custom U-Boot with NMBM. If firmware doesn't support NMBM but U-Boot support it, and a user flash his device with that firmware, it could potentially cause some issues. As a result, i add a specific layout.

ptpt52 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -35,7 +35,16 @@ redmi_ax6000_initial_setup()
+        local board=$(board_name)
+        case "$board" in
+        xiaomi,mi-router-wr30u-stock)
+                fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k>
```
what is this line?

hank9999 replied:
It may occurred during a replication. It has been corrected. Thanks for your review!

ptpt52 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -35,7 +35,16 @@ redmi_ax6000_initial_setup()
+                fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k>
+                ;;
+        xiaomi,redmi-router-ax6000-stock)
+                fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k>
```
1. may be wrong here.
2. use tab indent.

hank9999 replied:
Tab Corrected.

hank9999 replied:
mtdparts Corrected.

ptpt52 commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u-112m-nmbm.dts:
```diff
@@ -0,0 +1,22 @@
+&spi_nand {
+        mediatek,nmbm;
+        mediatek,bmt-max-ratio = <1>;
+        mediatek,bmt-max-reserved-blocks = <64>;
```
use tab instead of space here (3 lines)

hank9999 replied:
Corrected.

ptpt52 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -20,6 +20,11 @@ netgear,wax220)
+xiaomi,mi-router-wr30u-stock|\
+xiaomi,mi-router-wr30u-ubootmod)
+        ucidef_set_led_netdev "wan" "wan" "blue:network" "wan"
+        ;;
```
use tabs instead of spaces (2 lines)

hank9999 replied:
Corrected.

ptpt52 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -35,7 +35,16 @@ redmi_ax6000_initial_setup()
+        local board=$(board_name)
+        case "$board" in
+        xiaomi,mi-router-wr30u-stock)
+                fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k>
```
use tabs instead of spaces (4 lines)

hank9999 replied:
Corrected.

ptpt52 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -35,7 +35,16 @@ redmi_ax6000_initial_setup()
+        xiaomi,redmi-router-ax6000-stock)
+                fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k>
+                ;;
+        esac
```
use tabs instead of spaces (5 lines)

hank9999 replied:
Corrected.

ptpt52 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -357,6 +357,68 @@ define Device/tplink_tl-xdr6088
+  KERNEL_INITRAMFS := kernel-bin | lzma | \
+        fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
+  IMAGE/sysupgrade.itb := append-kernel | \
+        fit gzip $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-static-with-rootfs | append-metadata
```
use tabs instead of spaces

hank9999 commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u-stock.dts:
```diff
@@ -0,0 +1,34 @@
+/dts-v1/;
+#include "mt7981b-xiaomi-mi-router-wr30u.dtsi"
+
+/ { 
```
Corrected.

hank9999 commented on target/linux/mediatek/dts/mt7981b-xiaomi-mi-router-wr30u-ubootmod.dts:
```diff
@@ -0,0 +1,16 @@
+/dts-v1/;
+#include "mt7981b-xiaomi-mi-router-wr30u.dtsi"
+
+/ { 
```
Corrected.

hank9999 committed:
mediatek: filogic: add support for Xiaomi WR30U

Hardware specification:
  SoC: MediaTek MT7981B 2x A53
  Flash: ESMT F50L1G41LB 128MB
  RAM: NT52B128M16JR-FL 256MB
  Ethernet: 4x 10/100/1000 Mbps
  Switch: MediaTek MT7531AE
  WiFi: MediaTek MT7976C
  Button: Reset, Mesh
  Power: DC 12V 1A

Flash instructions:

1. Get ssh access
   Check this link: https://forum.openwrt.org/t/openwrt-support-for-xiaomi-ax3000ne/153769/22

2. Backup import partitions
   ```
   dev:    size   erasesize  name
   mtd1: 00100000 00020000 "BL2"
   mtd2: 00040000 00020000 "Nvram"
   mtd3: 00040000 00020000 "Bdata"
   mtd4: 00200000 00020000 "Factory"
   mtd5: 00200000 00020000 "FIP"
   mtd8: 02200000 00020000 "ubi"
   mtd9: 02200000 00020000 "ubi1"
   mtd12: 00040000 00020000 "KF"

   ```

   Use these commands blow to backup your stock partitions.
   ```
   nanddump -f /tmp/BL2.bin /dev/mtd1
   nanddump -f /tmp/Nvram.bin /dev/mtd2
   nanddump -f /tmp/Bdata.bin /dev/mtd3
   nanddump -f /tmp/Factory.bin /dev/mtd4
   nanddump -f /tmp/FIP.bin /dev/mtd5
   nanddump -f /tmp/ubi.bin /dev/mtd8
   nanddump -f /tmp/KF.bin /dev/mtd12
   ```
   Then, transfer them to your computer via scp, netcat, tftp
   or others and keep them in a safe place.

3. Setup Nvram
   Get the current stock: `cat /proc/cmdline`

   If you find `firmware=0` or `mtd=ubi`, use these commands:
   ```
   nvram set boot_wait=on
   nvram set uart_en=1
   nvram set flag_boot_rootfs=1
   nvram set flag_last_success=1
   nvram set flag_boot_success=1
   nvram set flag_try_sys1_failed=0
   nvram set flag_try_sys2_failed=0
   nvram commit
   ```

   If you find `firmware=1` or `mtd=ubi1`, use these commands:
   ```
   nvram set boot_wait=on
   nvram set uart_en=1
   nvram set flag_boot_rootfs=0
   nvram set flag_last_success=0
   nvram set flag_boot_success=1
   nvram set flag_try_sys1_failed=0
   nvram set flag_try_sys2_failed=0
   nvram commit
   ```

4. Flash stock-initramfs-factory.ubi
   If you find `firmware=0` or `mtd=ubi`:
   `ubiformat /dev/mtd9 -y -f /tmp/stock-initramfs-factory.ubi`

   If you find `firmware=1` or `mtd=ubi1`:
   `ubiformat /dev/mtd8 -y -f /tmp/stock-initramfs-factory.ubi`

   Then reboot your router, it should boot to the openwrt
   initramfs system now.

5. Setup uboot-env
   Now it will be setup automatically in upgrade process,
   you can skip this step.

   If your `fw_setenv` did not work, you need run this command:
   `echo "/dev/mtd1 0x0 0x10000 0x20000" > /etc/fw_env.config`

   Then setup uboot-env:
   ```
   fw_setenv boot_wait on
   fw_setenv uart_en 1
   fw_setenv flag_boot_rootfs 0
   fw_setenv flag_last_success 1
   fw_setenv flag_boot_success 1
   fw_setenv flag_try_sys1_failed 8
   fw_setenv flag_try_sys2_failed 8
   fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),
   2048k(factory),2048k(fip),256k(crash),256k(crash_log),
   34816k(ubi),34816k(ubi1),32768k(overlay),12288k(data),256k(KF)"
   ```

6. Flash stock-squashfs-sysupgrade.bin
   Use shell command:
   `sysupgrade -n /tmp/stock-squashfs-sysupgrade.bin`
   Or go to luci web.

If you need to change to Openwrt U-Boot layout, do next. If you
do not need, please ignore it.

Change to OpenWrt U-Boot:

1. Flash ubootmod-initramfs-factory.ubi
   Check mtd partitions: `cat /proc/mtd`
   ```
   mtd7: 00040000 00020000 "KF"
   mtd8: 02200000 00020000 "ubi_kernel"
   mtd9: 04e00000 00020000 "ubi"
   ```

   Run following command:
   `ubiformat /dev/mtd8 -y -f /tmp/ubootmod-initramfs-factory.ubi`
   Then reboot your router, it should boot to the openwrt initramfs
   system now.

2. Check mtd again
   ```
   mtd7: 00040000 00020000 "KF"
   mtd8: 07000000 00020000 "ubi"
   ```
   Make sure mtd8 is ubi.

3. Install kmod-mtd-rw
   Run command: `opkg update && opkg install kmod-mtd-rw`
   Or get it in openwrt server, or build it yourself, then install
   it manually

   Then run this command:
   `insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1`

4. Clean up pstore
   Run Command: `rm -f /sys/fs/pstore/*`

5. Format ubi and create new ubootenv volume
   ```
   ubidetach -p /dev/mtd8; ubiformat /dev/mtd8 -y; ubiattach -p /dev/mtd8
   ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
   ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB
   ```

6. (Optional) Add recovery boot feature.
   ```
   ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
   ubiupdatevol /dev/ubi0_2 /tmp/ubootmod-initramfs-recovery.itb
   ```

7. Flash Openwrt U-Boot
   ```
   mtd write /tmp/ubootmod-preloader.bin BL2
   mtd write /tmp/ubootmod-bl31-uboot.fip FIP
   ```

6. Flash ubootmod-squashfs-sysupgrade.itb
   Use shell command:
   `sysupgrade -n /tmp/ubootmod-squashfs-sysupgrade.itb`
   Or go to luci web.

Now everything is done, Enjoy!

Go Back to stock from Openwrt U-Boot:

1. Force flash ubootmod-initramfs-recovery.itb
   Use shell command:
   `sysupgrade -F -n /tmp/ubootmod-initramfs-recovery.itb`
   Or go to luci web.
   Then it should boot to the openwrt initramfs system now.

2. Format ubi and Nvram
   ```
   ubidetach -p /dev/mtd8; ubiformat /dev/mtd8 -y; ubiattach -p /dev/mtd8
   mtd erase Nvram
   ```

3. Install kmod-mtd-rw
   Run command: `opkg update && opkg install kmod-mtd-rw`
   Or get it in openwrt server, or build it yourself, then install
   it manually

   Then run this command:
   `insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1`

4. Flash stock U-Boot and ubi
   ```
   mtd write /tmp/BL2.bin BL2
   mtd write /tmp/FIP.bin FIP
   mtd write /tmp/ubi.bin ubi
   ```
   Then reboot your router, waiting it finished rollback in minutes.

Go Back to stock from stock layout Openwrt:
   Just run command: `ubiformat /dev/mtd8 -y -f /tmp/ubi.bin`
   Then reboot your router, waiting it finished rollback in minutes.

Notes:
1. Openwrt U-Boot and ubootmod openwrt did not enable NMBM.
   Please make your backup safe.

Signed-off-by: Hank Moretti <mchank9999@gmail.com>

hauke commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -47,6 +47,7 @@ mercusys,mr90x-v1)
 	ubootenv_add_uci_config "/dev/mtd5" "0x0" "0x20000" "0x20000"
 	;;
+xiaomi,mi-router-wr30u-stock|\
 xiaomi,redmi-router-ax6000-stock)
```
You have to handle `xiaomi,mi-router-wr30u-112m-nmbm` here too.

hank9999 replied:
Corrected. Thank you.

hank9999 committed:
mediatek: filogic: add specific layout for WR30U

Because this device enable NMBM by default, most users use custom
U-Boot with NMBM-Enabled in Chinese forums.

This layout is the same as the ubootmod layout but enabling NMBM.

Signed-off-by: Hank Moretti <mchank9999@gmail.com>

hank9999 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -79,10 +87,13 @@ platform_do_upgrade() {
 	tplink,tl-xdr4288|\
 	tplink,tl-xdr6086|\
 	tplink,tl-xdr6088|\
+	xiaomi,mi-router-wr30u-112m-nmbm|\
```
that's right, it has a wrong here.
but, how can i fix it, another pr?

hank9999 commented:
> **Are Xiaomi router wr30u and Xiaomi router ax3000t compatible?** Compared with WR30U, Xiaomi router AX3000T has one less LED light and more NFC function. AX3000T Purchase link: https://www.mi.com/xiaomi-ax3000t It is said that there is no elink service in the firmware of AX3000T, so the vulnerability of wr30u to unlock ssh cannot be used anymore. If ssh can be unlocked later, can I directly flash the official openwrt firmware of WR30U? If not, after adding AX3000T device support in the future, will it be able to achieve the same NFC function as Xiaomi’s official firmware?

Yes, only the LED and NFC functions make a difference. Someone tried flashing wr30u with ax3000t firmware, and it works fine without the LED and NFC. The LED problem can easily be fixed by modifying the dts file, but I don't know how the NFC function works. I heard that a new way to unlock ssh has been found, but lean said it will be published after 11.11.

hank9999 commented:
> > > **Are Xiaomi router wr30u and Xiaomi router ax3000t compatible?** Compared with WR30U, Xiaomi router AX3000T has one less LED light and more NFC function. AX3000T Purchase link: https://www.mi.com/xiaomi-ax3000t It is said that there is no elink service in the firmware of AX3000T, so the vulnerability of wr30u to unlock ssh cannot be used anymore. If ssh can be unlocked later, can I directly flash the official openwrt firmware of WR30U? If not, after adding AX3000T device support in the future, will it be able to achieve the same NFC function as Xiaomi’s official firmware?
> > 
> > Yes, only the LED and NFC functions make a difference. Someone tried flashing wr30u with ax3000t firmware, and it works fine without the LED and NFC. The LED problem can easily be fixed by modifying the dts file, but I don't know how the NFC function works. I heard that a new way to unlock ssh has been found, but lean said it will be published after 11.11.
> 
> Okay, but wr30u is an ISP customized version and ax3000t is a retail version.
> So will wr30u be renamed to ax3000t and be compatible with wr30u? Or add another device to support ax3000t?

I think it is better to add a new device, although they have the similar hardware, but they are too device in market.

upsangel commented:
Thank you @hank9999 @aiamadeus @ptpt52 making this great router OpenWRT ready. 
I use the luci-gui "Save mtdblock contents" to save mtd9 to a bin file, then recover by mtd write the bin file to ubi. Result: brick...
Now the WR30U enters boot loops, but **still can enter "emergency mode**" if I long press the reset button ( 系統 long orange light, then short blinks for few seconds, then repeat)

May I ask what is the best way now to recover the system? I followed the guide to flash OpenWRT so I guess the bootloader/uboot should be intact? I am wondering if the TFTP method is available if so what will be the IP setting to "feed" the initramfs?

Thank you very much in advance!

hank9999 commented:
> Thank you @hank9999 @aiamadeus @ptpt52 making this great router OpenWRT ready. I follow steps by steps and everything works.
> 
> I use the luci-gui "Save mtdblock contents" to save mtd9 to a bin file, then recover by `mtd write` the bin file to ubi. Result: brick... Now the WR30U enters boot loops, but **still can enter "emergency mode**" if I long press the reset button ( 系統 long orange light, then short blinks for few seconds, then repeat)
> 
> May I ask what is the best way now to recover the system? I followed the guide to flash OpenWRT so I guess the bootloader/uboot should be intact? I am wondering if the TFTP method is available if so what will be the IP setting to "feed" the initramfs?
> 
> Thank you very much in advance!

Hi @upsangel , I am sorry to heard that your router was in brick. Did you use OpenWRT u-boot or u-boot with failsafe web?

upsangel commented:
> Hi @upsangel , I am sorry to heard that your router was in brick. Did you use OpenWRT u-boot or u-boot with failsafe web?

I guess I am still on "stock u-boot", since the only flash commands I use were:
ubiformat /dev/mtd9 -y -f /tmp/stock-initramfs-factory.ubi
and
sysupgrade -n /tmp/stock-squashfs-sysupgrade.bin

And that's it, I haven't flash or touch u-boot layout etc.

hank9999 commented:
> > Hi @upsangel , I am sorry to heard that your router was in brick. Did you use OpenWRT u-boot or u-boot with failsafe web?
> 
> Wow you reply so quick! Amazing!!!
> 
> I guess I am still on "stock u-boot", since the only flash commands I use were: ubiformat /dev/mtd9 -y -f /tmp/stock-initramfs-factory.ubi and sysupgrade -n /tmp/stock-squashfs-sysupgrade.bin
> 
> And that's it, I haven't flashed or touched u-boot layout etc.
> 
> Thank you!

@upsangel I just sat down and look at my computer. I think you may need to use ttl to unbrick your device. 
```
fw_setenv boot_wait on
fw_setenv uart_en 1
```
These env will keep uboot function online. Connect to TTL 115200
Then use any key to pause it. I think down key is good idea. After entering `U-Boot Boot Menu`, use `Upgrade ATF FIP`, use TFTP to  upload a u-boot with failsafe (such as https://github.com/hanwckf/bl-mt798x) or Openwrt U-Boot will make it easy.

upsangel commented:
Thank you for your advice! I will try TTL to flash hanwckf/bl-mt798x as it has a nice web gui. 

Just come across this: https://schaepher.github.io/2019/10/07/xiaomi-router-r3-recovery/ he suggests to input below in the TTL console:

```
setenv boot_wait on
setenv uart_en 1
saveenv
```

Not sure I should follow the above or as what you suggested: 

> ```
> fw_setenv boot_wait on
> fw_setenv uart_en 1
> ```
> 

Hmm... Excuse my lack of TTL experience...

hank9999 commented:
> Thank you for your advice! I will try TTL to flash hanwckf/bl-mt798x as it has a nice web gui.
> 
> Just come across this: https://schaepher.github.io/2019/10/07/xiaomi-router-r3-recovery/ he suggests to input below in the TTL console:
> 
> ```
> setenv boot_wait on
> setenv uart_en 1
> saveenv
> ```
> 
> Not sure I should follow the above or as what you suggested:
> 
> > ```
> > fw_setenv boot_wait on
> > fw_setenv uart_en 1
> > ```
> 
> Hmm... Excuse my lack of TTL experience...

Hmm, sorry, I may not have expressed it clearly. I mean it was been executed in install progress, so you can just use ttl to flash it.

hank9999 commented:
> Thank you for your advice! I will try TTL to flash hanwckf/bl-mt798x as it has a nice web gui.
> 
> Just come across this: https://schaepher.github.io/2019/10/07/xiaomi-router-r3-recovery/ he suggests to input below in the TTL console:
> 
> ```
> setenv boot_wait on
> setenv uart_en 1
> saveenv
> ```
> 
> Not sure I should follow the above or as what you suggested:
> 
> > ```
> > fw_setenv boot_wait on
> > fw_setenv uart_en 1
> > ```
> 
> Hmm... Excuse my lack of TTL experience...

1. You need to setup tftp and put file into the same folder
2. Connect to ttl then connect power
3. Press down until enter `U-Boot Boot Menu`, select `Upgrade ATF FIP`
4. Input some address and file name, done.

No more other actions.

upsangel commented:
> No more other actions.

Thank you! I will give it a try, if successful I will make a write up / gist of take away. Update later. Cheers,

hank9999 commented:
> > No more other actions.
> 
> Thank you! I will give it a try, if successful I will make a write up / gist of take away. Update later. Cheers,

Thank you!

upsangel commented:
> > Thank you for your advice! I will try TTL to flash hanwckf/bl-mt798x as it has a nice web gui.
> > Just come across this: https://schaepher.github.io/2019/10/07/xiaomi-router-r3-recovery/ he suggests to input below in the TTL console:
> > ```
> > setenv boot_wait on
> > setenv uart_en 1
> > saveenv
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure I should follow the above or as what you suggested:
> > > ```
> > > fw_setenv boot_wait on
> > > fw_setenv uart_en 1
> > > ```
> > 
> > 
> > Hmm... Excuse my lack of TTL experience...
> 
> 1. You need to setup tftp and put file into the same folder
> 2. Connect to ttl then connect power
> 3. Press down key until enter `U-Boot Boot Menu`, select `Upgrade ATF FIP`
> 4. Input some address and file name, done. It will get image from tftp server in your computer.
> 5. Reboot with press reset key, it will enter web failsafe
> 
> No more other actions.

Hi @hank9999 , glad to let you know that your instruction works like a charm, one kick and one goal
supplement two minor things for others reference:
1. the TTL jumper sequence https://www.right.com.cn/forum/thread-8310997-2-1.html
2. I used the mt7981_wr30u-fip-fixed-parts-multi-layout.bin from https://github.com/hanwckf/bl-mt798x

Much appreciate for your help again!

Just one more question if you don't mind, I saw in the MTK's uboot menu, there is a option called:  `Upgrade firmware` I am wondering what "firmware" is referred to? Is firmware = ubi partition? Or firmware = dd dump of all the partitions combined? Thank you!

hank9999 commented:
> > > Thank you for your advice! I will try TTL to flash hanwckf/bl-mt798x as it has a nice web gui.
> > > Just come across this: https://schaepher.github.io/2019/10/07/xiaomi-router-r3-recovery/ he suggests to input below in the TTL console:
> > > ```
> > > setenv boot_wait on
> > > setenv uart_en 1
> > > saveenv
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Not sure I should follow the above or as what you suggested:
> > > > ```
> > > > fw_setenv boot_wait on
> > > > fw_setenv uart_en 1
> > > > ```
> > > 
> > > 
> > > Hmm... Excuse my lack of TTL experience...
> > 
> > 
> > 
> > 1. You need to setup tftp and put file into the same folder
> > 2. Connect to ttl then connect power
> > 3. Press down key until enter `U-Boot Boot Menu`, select `Upgrade ATF FIP`
> > 4. Input some address and file name, done. It will get image from tftp server in your computer.
> > 5. Reboot with press reset key, it will enter web failsafe
> > 
> > No more other actions.
> 
> Hi @hank9999 , glad to let you know that your instruction works like a charm, one kick and one goal supplement two minor things for others reference:
> 
> 1. the TTL jumper sequence https://www.right.com.cn/forum/thread-8310997-2-1.html
> 2. I used the mt7981_wr30u-fip-fixed-parts-multi-layout.bin from https://github.com/hanwckf/bl-mt798x
> 
> Much appreciate for your help again!
> 
> Just one more question if you don't mind, I saw in the MTK's uboot menu, there is a option called: `Upgrade firmware` I am wondering what "firmware" is referred to? Is firmware = ubi partition? Or firmware = dd dump of all the partitions combined? Thank you!

According to the [code1](https://github.com/hanwckf/bl-mt798x/blob/af13dbd3cc474acc1d781d9dc88b00299fb02142/uboot-mtk-20220606/board/mediatek/mt7981/bootmenu_ubi.c#L57C13-L57C13) and [code2](https://github.com/hanwckf/bl-mt798x/blob/af13dbd3cc474acc1d781d9dc88b00299fb02142/uboot-mtk-20220606/board/mediatek/common/ubi_helper.c#L537), it will upgrade ubi
```
ret = ubi_upgrade_image(data, size);
```

hank9999 commented:
> > > redmi_ax6000 has implemented this button: `Mesh = BTN_9` What should I do if I also want to implement it on wr30u?
> > 
> > 
> > In the dts file `gpio-keys`, wr30u have also configured the mesh button as btn 9.
> 
> `mesh { label = "mesh"; gpios = <&pio 102 GPIO_ACTIVE_LOW>; linux,code = <BTN_9>; linux,input-type = <EV_SW>; };` Is that so? How can I know the gpios of the wr30u mesh button?

yes, they have the same mesh button gpio config. The way to test gpio can be found in openwrt wiki or you can make deal with the firmware to get dts. In other project, some added ax3000t support you can check them, for example https://github.com/hanwckf/immortalwrt-mt798x/pull/160

hank9999 commented:
> > > > > redmi_ax6000 has implemented this button: `Mesh = BTN_9` What should I do if I also want to implement it on wr30u?
> > > > 
> > > > 
> > > > In the dts file `gpio-keys`, wr30u have also configured the mesh button as btn 9.
> > > 
> > > 
> > > `mesh { label = "mesh"; gpios = <&pio 102 GPIO_ACTIVE_LOW>; linux,code = <BTN_9>; linux,input-type = <EV_SW>; };` Is that so? How can I know the gpios of the wr30u mesh button?
> > 
> > 
> > yes, they have the same mesh button gpio config. The way to test gpio can be found in openwrt wiki or you can make deal with the firmware to get dts. In other project, some added ax3000t support you can check them, for example [hanwckf/immortalwrt-mt798x#160](https://github.com/hanwckf/immortalwrt-mt798x/pull/160)
> 
> Flashing wr30u-stock-sysupgrade.bin and wr30u-112m-nmbm-factory.ubi on ax3000t can work normally, but when upgrading wr30u-112m-nmbm-sysupgrade.bin from wr30u-112m-nmbm-factory.ubi, the yellow light keeps on Cannot start normally

What uboot you are using? stock, OpenWRT U-Boot or bl-mt798x?

hank9999 commented:
> > stock
> 
> stock

stock uboot can only support stock layout firmware. if you want to use 112m-nmbm please use bl-mt798x project. if you want to use ubootmod, please use OpenWRT uboot and preloader.

hank9999 commented:
> Why does the bl-mt798x project only need to flash uboot.fip, but openwrt also needs to flash preloader.bin? What is the file preloader.bin? If I don't flash preloader.bin, can uboot start normally?

Actually, I did not know it clearly, I am new in this area. preloader.bin is BL2, and I did not try OpenWRT U-Boot without flashing preloader.

upsangel commented:
> redmi_ax6000 has implemented this button: `Mesh = BTN_9` What should I do if I also want to implement it on wr30u?

Hi friend, I share my experience customizing this button: https://upsangel.com/openwrt/openwrt%e8%87%aa%e5%ae%9a%e7%be%a9%e6%a9%9f%e8%ba%ab%e6%8c%89%e9%8d%b5%e5%8a%9f%e8%83%bd%e6%95%99%e5%ad%b8%ef%bc%9a%e7%88%b2wr30u%e6%b7%bb%e5%8a%a0wifi%e9%96%8b%e9%97%9c%e6%8c%89%e9%8d%b5/

It's in Chinese, I hope ChatGPT / Google translate will help you turn it into your language if needed. Cheers,

axyzs commented:
> > Why does the bl-mt798x project only need to flash uboot.fip, but openwrt also needs to flash preloader.bin? What is the file preloader.bin? If I don't flash preloader.bin, can uboot start normally?
> 
> Actually, I did not know it clearly, I am new in this area. preloader.bin is BL2, and I did not try OpenWRT U-Boot without flashing preloader.

Hi @hank9999 thanks for the amazing job to support WR30U. I flashed [uboot.fip](https://firmware-selector.openwrt.org/?version=23.05.2&target=mediatek%2Ffilogic&id=xiaomi_mi-router-wr30u-ubootmod) from the official download link but without flashing preloader.bin. I assumed that the official uboot.fip didn't require preloader.bin. Now, I am unable to see the uboot web UI and cannot SSH into the device. Does this mean I bricked the device? I backed up each partition. Is there any way to restore the device to working condition? Would flashing via the serial port work? Thanks.

dangowrt commented:
@axyzs preloader from stock firmware is incompatible with the way ARM Trusted Firmware-A and U-Boot are built in more recent MediaTek SDK builds and also in OpenWrt.

At this point the only options you have are:
 - physically remove the flash chip from the board, write to it using and external SPI flashing tool, place it back on the board.
 - try locating JTAG pins and load U-Boot into RAM via JTAG, then use this to re-write the SPI flash.

MuSCII commented:
Many thanks @dangowrt for your advice. It looks like it has to be fixed by touching the hardware.

axyzs commented:
Thanks a lot @dangowrt

hank9999 commented:
> Thanks a lot @dangowrt 

@axyzs I have checked the PCB image, I am not sure if Xiaomi make the JTAG connector public. You can try to flash the physical flash chip with a programmer, such as CH341A. CH341A has enough function with cheap price.

axyzs commented:
Many thanks @hank9999 looks either way needs a hardware to help fixing the issue.

axyzs commented:
Hi @hank9999, I have a question since I've never used a programmer before. If I were able to flash the chip using a CH341A programmer, which file should be flashed? I followed your steps and backed up all the partitions as below

```
nanddump -f /tmp/BL2.bin /dev/mtd1
nanddump -f /tmp/Nvram.bin /dev/mtd2
nanddump -f /tmp/Bdata.bin /dev/mtd3
nanddump -f /tmp/Factory.bin /dev/mtd4
nanddump -f /tmp/FIP.bin /dev/mtd5
nanddump -f /tmp/ubi.bin /dev/mtd8
nanddump -f /tmp/KF.bin /dev/mtd12
```

hank9999 commented:
@axyzs Hi, The programmer need to flash the full flash chip. So you need to read the flash first, the use hex editor to replace the uboot data, next, write it back. 

Or, you can just flash the a full flash  backup from the Internet, then flush the bdata and factory partitions back.

axyzs commented:
Thanks for your reply @hank9999

axyzs commented:
For those people like me who only flashed the OpenWRT official uboot.fip from the official download link but did not flash preloader.bin, this is just an idea or perhaps a potential solution. **I CANNOT GUARANTEE it would work** as I failed, but I flashed a third-party file to FIP! If you don't have a programmer, it might be worth a try!

Obviously, after flashing only the FIP, you won't be able to connect the device via the network, so TFTP will not work at all because the Ethernet ports won't function. I assume this is because OpenWRT doesn't contain the device's network driver information, preventing the network from loading (though I could be wrong). However, I tried all four ports, and none of them worked, neither 192.168.1.1 nor 192.168.31.1 since the network could not be started or booted.

It was not successful for me because I didn't flash the original FIP file that I had backed up. I struggled to flash the original FIP or the file from the bl-mt798x project. I chose the latter one, but it didn't work. It had nothing to do with the file from the bl-mt798x project, I may have done something wrong! 

Before commencing this solution, ensure you have two necessary items:

- A USB to TTL Adapter for connecting the device using TTL.
- The original FIP file that you backed up.

There are four holes on the board, from top to bottom: ACC, RX, Ground, TX. You don't need to connect ACC, just connect RX, Ground, and TX to your USB to TTL Adapter. If you connect it correctly, you will see the OpenWRT uboot menu. Choose the U-Boot console.

Use the loadx command to enable file transfer using xmodem. You could try ExtraPuTTY if you are using Windows.

This is what I got: the file was transferred to the RAM.

```
MT7981> loadx
## Ready for binary (xmodem) download to 0x46000000 at 115200 bps...
CCC## Total Size      = 0x000b3861 = 735329 Bytes
## Start Addr      = 0x46000000
```

then run `mtd list` to find the the value of "fip" 

```
MT7981> mtd list
List of MTD devices:
* spi-nand0
  - device: spi_nand@0
  - parent: spi@1100a000
  - driver: spi_nand
  - path: /spi@1100a000/spi_nand@0
  - type: NAND flash
  - block size: 0x20000 bytes
  - min I/O: 0x800 bytes
  - OOB size: 64 bytes
  - OOB available: 24 bytes
  - 0x000000000000-0x000008000000 : "spi-nand0"
          - 0x000000000000-0x000000100000 : "bl2"
          - 0x000000100000-0x000000140000 : "Nvram"
          - 0x000000140000-0x000000180000 : "Bdata"
          - 0x000000180000-0x000000380000 : "factory"
          - 0x000000380000-0x000000580000 : "fip"
          - 0x000000580000-0x0000005c0000 : "crash"
          - 0x0000005c0000-0x000000600000 : "crash_log"
          - 0x000000600000-0x000007600000 : "ubi"
          - 0x000007600000-0x000007640000 : "KF"
```

then I did 

```
MT7981> nand write 0x46000000 0x380000 735329
```

0x46000000 to match `## Start Addr`
0x380000 to match the start value of FIP: 0x000000380000-0x000000580000
735329 to match the file size. 



After rebooting my device, it got stuck at 'BL2: Booting BL31.' and stopped to continue. I don't know what would happen if I flash the original FIP, but it seems that what I did was partially correct. So please try the original FIP. 

```
NOTICE:  EMI: Detected DRAM size: 256MB
NOTICE:  EMI: complex R/W mem test passed
NOTICE:  CPU: MT7981 (1300MHz)
NOTICE:  SPI_NAND parses attributes from parameter page.
NOTICE:  SPI_NAND Detected ID 0xc8
NOTICE:  Page size 2048, Block size 131072, size 134217728
NOTICE:  Initializing NMBM ...
NOTICE:  Signature found at block 1023 [0x07fe0000]
NOTICE:  First info table with writecount 0 found in block 960
NOTICE:  Second info table with writecount 0 found in block 963
NOTICE:  NMBM has been successfully attached in read-only mode
NOTICE:  BL2: Booting BL31
```

If the above steps didn't work for you, ultimately, you will need a programmer to help. That could be the only way.

Now, I need a bit of help. If anyone knows where to find the full flash backup, I've searched for quite a while but couldn't find anywhere that provides the full flash backup. Or are there any projects that provide that file?  Thanks in advanced.

hank9999 commented:
@axyzs I upload the backup files that have removed sensitive data.  [hank9999/WR30U-Backups](https://github.com/hank9999/WR30U-Backups)

axyzs commented:
Many thanks @hank9999 I appreciate your help!

hank9999 commented:
> > > > Why does the bl-mt798x project only need to flash uboot.fip, but openwrt also needs to flash preloader.bin? What is the file preloader.bin? If I don't flash preloader.bin, can uboot start normally?
> > > 
> > > 
> > > Actually, I did not know it clearly, I am new in this area. preloader.bin is BL2, and I did not try OpenWRT U-Boot without flashing preloader.
> > 
> > 
> > Hi @hank9999 thanks for the amazing job supporting WR30U Device. I flashed [uboot.fip](https://firmware-selector.openwrt.org/?version=23.05.2&target=mediatek%2Ffilogic&id=xiaomi_mi-router-wr30u-ubootmod) from the official download link but without flashing preloader.bin. I assumed that the official uboot.fip didn't require preloader.bin. Now, I am unable to see the uboot web UI and cannot SSH into the device. Does this mean I bricked the device? I backed up each partition. Is there any way to restore the device to working condition? Would flashing via the serial port work? Thanks.
> 
> flashed uboot.fip from the official openwrt not have web UI

the official openwrt uboot do not have web failsafe, it's failsafe is based on tty, inside recovery system and tftp boot. web failsafe is from the bl-mt798x project

Miigon commented:
@axyzs Your idea should've worked, especially if you have the stock partition backups.
However in your case you most likely failed because you forgot to `mtd erase` or `nand erase` before flashing the BL. Recall that nand flash bits can only be changed from 1 to 0 and never 0 to 1. You need to erase the blocks before programming the nand, or else you end up with the bit-AND of the old and new bootloaders in nand, which is a garbled mess.

h9419 commented:
@hank9999 I made the same mistake as @axyzs and bricked my WR30U. 

I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 

Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 

Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?

hank9999 commented:
> @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> 
> I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> 
> Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> 
> Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?

Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.

Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.

h9419 commented:
> > @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> > 
> > I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> > 
> > Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> > 
> > Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?
> 
> Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.
> 
> Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.
> 

I was trying to use flashrom with pico-serprog on a pi-pico (RP2040). It could be flashrom not supporting the memory map

What software/hardware did you use to dump the F50L1G41LB ? I can borrow a USB ISP but I don't know how to use that. I am more comfortable with the pico-serprog since it is 3V3 and documented by libreboot project

hank9999 commented:
> > > @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> > > 
> > > I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> > > 
> > > Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> > > 
> > > Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?
> > 
> > Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.
> > 
> > Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.
> > 
> 
> I was trying to use flashrom with pico-serprog on a pi-pico (RP2040). It could be flashrom not supporting the memory map
> 
> What software/hardware did you use to dump the F50L1G41LB ? I can borrow a USB ISP but I don't know how to use that. I am more comfortable with the pico-serprog since it is 3V3 and documented by libreboot project 

Hi, I am checking the flashroom document,   finding that it seems that it has not support nand flash widely, and there is a pull request to add support it, try using this pr.

I am using CH341A with NeoProgrammer to flash F50L1G41LB chips, it is slow but working.

hank9999 commented:
> > > @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> > > 
> > > I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> > > 
> > > Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> > > 
> > > Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?
> > 
> > Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.
> > 
> > Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.
> > 
> 
> I was trying to use flashrom with pico-serprog on a pi-pico (RP2040). It could be flashrom not supporting the memory map
> 
> What software/hardware did you use to dump the F50L1G41LB ? I can borrow a USB ISP but I don't know how to use that. I am more comfortable with the pico-serprog since it is 3V3 and documented by libreboot project 

Could you give me more Information about that your borrowed ISP,  I'll check what is.

h9419 commented:
> > > > @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> > > > 
> > > > I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> > > > 
> > > > Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> > > > 
> > > > Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?
> > > 
> > > Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.
> > > 
> > > Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.
> > > 
> > 
> > I was trying to use flashrom with pico-serprog on a pi-pico (RP2040). It could be flashrom not supporting the memory map
> > 
> > What software/hardware did you use to dump the F50L1G41LB ? I can borrow a USB ISP but I don't know how to use that. I am more comfortable with the pico-serprog since it is 3V3 and documented by libreboot project 
> 
> Could you give me more Information about that your borrowed ISP,  I'll check what is.
> 
> 

I don't have it with me right now. It'll likely be another week until I go back and try again. I will write back once I have tried different software/hardware. 

Thank you for maintaining this and pointing out that NAND is different from what I'm familiar with. I'm considering buying a CH341A. If you have a known good taobao link I would appreciate it. Still worried about the 3V3/5V issue

hank9999 commented:
> > > > > @hank9999 I made the same mistake as @axyzs and bricked my WR30U. 
> > > > > 
> > > > > I have soldered out the flash chip but flashrom does not recognize it. I have a 64MB winbond chip also in WSON8 package that can be read and written to using the same programmer. 
> > > > > 
> > > > > Does it mean my 128MB SPI flash chip I desoldered from the router is toast? I don't have another 128MB flash but I do have a 64MB winbond. 
> > > > > 
> > > > > Is it possible to make a custom build of openWRT, write it to my 64MB `W25Q512JV` and unbrick my router with it?
> > > > 
> > > > Generally speaking, this NAND SPI flash chip is not that easy to damage by just soldering it.
> > > > 
> > > > Try checking the connection, and also check if the software supports it. If you are using a CH341A, please try NeoProgrammer.
> > > > 
> > > 
> > > I was trying to use flashrom with pico-serprog on a pi-pico (RP2040). It could be flashrom not supporting the memory map
> > > 
> > > What software/hardware did you use to dump the F50L1G41LB ? I can borrow a USB ISP but I don't know how to use that. I am more comfortable with the pico-serprog since it is 3V3 and documented by libreboot project 
> > 
> > Could you give me more Information about that your borrowed ISP,  I'll check what is.
> > 
> > 
> 
> I don't have it with me right now. It'll likely be another week until I go back and try again. I will write back once I have tried different software/hardware. 
> 
> Thank you for maintaining this and pointing out that NAND is different from what I'm familiar with. I'm considering buying a CH341A. If you have a known good taobao link I would appreciate it. Still worried about the 3V3/5V issue

Sorry about that, my ch341a programmer was bought six years ago, I just buy a random one, and it works. But, my programmer's output voltage is 3.3V by default. It was a dizzy.

hank9999 commented:
@h9419 There is a project, probably a good choice for you, https://github.com/981213/spi-nand-prog
It support programming nand flash with serprog
Another is AsProgrammer, its newest version support flash by a arduino serial port.

h9419 commented:
@hank9999 Thanks for your direction. 


`ufprog` had little to no documentation on using serprog and it didn't do much for me, and  `spi-nand-prog` didn't support any ESNT NAND. 


However, according to `SNANDer` source code, the layout is similar enough between `F50L1G41LB` and `W25N01G`. I managed to modify `spi-nand-prog` enough that it recognizes the layout of the NAND and dumped out the flash. Verified that it looks similar to what your backup looks like and DD your FIP partition over my dump then flash it back.


Unfortunately, now it flashes orange around 14 times periodically instead of constant orange. It's a little late and I will try again later

hank9999 commented:
> > > @hank9999 Thanks for your direction. 
> > > `ufprog` had little to no documentation on using serprog and it didn't do much for me, and  `spi-nand-prog` didn't support any ESNT NAND. 
> > > However, according to `SNANDer` source code, the layout is similar enough between `F50L1G41LB` and `W25N01G`. I managed to modify `spi-nand-prog` enough that it recognizes the layout of the NAND and dumped out the flash. Verified that it looks similar to what your backup looks like and DD your FIP partition over my dump then flash it back.
> > > Unfortunately, now it flashes orange around 14 times periodically instead of constant orange. It's a little late and I will try again later
> > 
> > 
> > @h9419   I have some idea for you.  Try to do a full nand backup. After that use a hex editor to check your bdata and factory partitions, searching as mac, ssid, mi iot token, sn in bdata, and wifi mac, wifi calibration in factory. If they all exist, it proved that your programmer works well,  And keep them in a safe place, it will be used later.
> > Then write my full nand backup and check if your device can boot.  If it boot into the system, you can use ssh tool to enable ssh, and write back your bdata and factory, your device will reborn.
> 
> Thanks. I looked at my dump and the things you mentioned are safely in the Bdata partition. I am not sure on how to parse the factory partition.
> 
> It's nerve-wracking to have to solder that SMT out again tho

You can check you wifi mac in factory partition, 0x04-0x07 and 0x0A-0x0D should be same as your wifi mac 4 bytes

h9419 commented:
> > > > @hank9999 Thanks for your direction. 
> > > > `ufprog` had little to no documentation on using serprog and it didn't do much for me, and  `spi-nand-prog` didn't support any ESNT NAND. 
> > > > However, according to `SNANDer` source code, the layout is similar enough between `F50L1G41LB` and `W25N01G`. I managed to modify `spi-nand-prog` enough that it recognizes the layout of the NAND and dumped out the flash. Verified that it looks similar to what your backup looks like and DD your FIP partition over my dump then flash it back.
> > > > Unfortunately, now it flashes orange around 14 times periodically instead of constant orange. It's a little late and I will try again later
> > > 
> > > 
> > > @h9419   I have some idea for you.  Try to do a full nand backup. After that use a hex editor to check your bdata and factory partitions, searching as mac, ssid, mi iot token, sn in bdata, and wifi mac, wifi calibration in factory. If they all exist, it proved that your programmer works well,  And keep them in a safe place, it will be used later.
> > > Then write my full nand backup and check if your device can boot.  If it boot into the system, you can use ssh tool to enable ssh, and write back your bdata and factory, your device will reborn.
> > 
> > 
> > Thanks. I looked at my dump and the things you mentioned are safely in the Bdata partition. I am not sure on how to parse the factory partition.
> > It's nerve-wracking to have to solder that SMT out again tho
> 
> You can check you wifi mac in factory partition, 0x04-0x09 and 0x0A-0x0F should be same as your wifi mac (2.4G and 5G)

Thanks, it checks out. 

I will try again with my dump for the first 3.5 MiB (BL2, Nvram, Bdata & Factory) and use your backup from FIP onwards. As I understand, FIP is for u-boot, ubi is the OS, what does BL2 stands for?

hank9999 commented:
Basically, the datas in nand are all same except Bdata, Factory, Nvram (some personal settings), crash (a log partition)

h9419 commented:
@hank9999 Thank you very much. I got it back to working using the following commands. Writing it down for future reference 

``` bash
cp dump.bin ok.bin
# replace content if ok.bin from 3.5MB onwards  with backup
# this creates an image with my own nvram, bdata and factory but with a known good uboot and ubi
dd if=wr30u-full-flash-backup.bin bs=524288 seek=7 skip=7 of=ok.bin
```

Then flash `ok.bin` and it boots as if I had a factory new router. I installed OpenWRT immediately but with stock layout.

hank9999 commented:
> @hank9999 Thank you very much. I got it back to working using the following commands. Writing it down for future reference
> 
> ```shell
> cp dump.bin ok.bin
> # replace content if ok.bin from 3.5MB onwards  with backup
> # this creates an image with my own nvram, bdata and factory but with a known good uboot and ubi
> dd if=wr30u-full-flash-backup.bin bs=524288 seek=7 skip=7 of=ok.bin
> ```
> 
> Then flash `ok.bin` and it boots as if I had a factory new router. I installed OpenWRT immediately but with stock layout.

I am glad to know that and wish you all the best.
